### PR TITLE
fix(ivy): ensure `ngClass` works with [class] bindings

### PR DESCRIPTION
**Pull request by matsko**

### CODE REVIEW

File: packages/core/src/render3/instructions.ts [hunk 1]
```diff
@@ -1543,13 +1536,22 @@ export function elementStyling(
...
-        createStylingContextTemplate(classDeclarations, styleDeclarations, styleSanitizer);
+    tNode.stylingTemplate = createStylingContextTemplate(
+        classDeclarations, styleDeclarations, styleSanitizer, !!dirWithClassInput);
   }
+
   if (styleDeclarations && styleDeclarations.length ||
       classDeclarations && classDeclarations.length) {
+    // directives with a [class] input expect the starting CSS classes to be passed in
+    if (dirWithClassInput) {
+      setInputsForProperty(
```
Comment (mhevery) [inline packages/core/src/render3/instructions.ts]: It seems to me that this well execute every single CD. There is no guard to not do this on CD if no change.

File: packages/core/src/render3/instructions.ts [hunk 2]
```diff
@@ -1642,7 +1644,12 @@ export function elementStyleProp(
 export function elementStylingMap<T>(
     index: number, classes: {[key: string]: any} | string | null,
     styles?: {[styleName: string]: any} | null): void {
-  updateStylingMap(getStylingContext(index, viewData), classes, styles);
+  const stylingContext = getStylingContext(index, viewData);
+  if (delegateToClassInput(stylingContext)) {
+    setInputsForProperty(getTNode(index, viewData).inputs !['class'] !, classes);
```
Comment (mhevery) [inline packages/core/src/render3/instructions.ts]: another megamorphic read here.

File: packages/core/test/render3/integration_spec.ts [hunk 3]
```diff
@@ -1517,6 +1518,77 @@ describe('render3 integration test', () => {
         expect(fixture.html)
             .toEqual('<structural-comp class="">Comp Content</structural-comp>Temp Content');
       });
+
+      it('should delegate all initial classes to a [class] input binding if present on a directive on the same element',
+         () => {
+           let mockClassDirective: DirWithClassDirective;
```
Comment (mhevery) [inline packages/core/test/render3/integration_spec.ts]: Could you refactor common code out to make it more readable? Also can you add a test to assert that we don't call `klass` setter if there is no change in the expression.

**Review by mhevery — CHANGES_REQUESTED** [review]
Some fixes here 2a23337cd0

File: packages/core/test/render3/integration_spec.ts [hunk 4]
```diff
@@ -1517,6 +1518,63 @@ describe('render3 integration test', () => {
...
+             }
+             if (rf & RenderFlags.Update) {
+               elementStylingMap(0, 'cucumber grape');
+               elementStylingApply(0);
+             }
+           }, 1, 0, [DirWithClassDirective]);
+
+           const fixture = new ComponentFixture(App);
+           expect(mockClassDirective !.classesVal).toEqual('cucumber grape');
+         });
```
Comment (pkozlowski-opensource) [inline packages/core/test/render3/integration_spec.ts]: I would love to see another test that mixes `[class]="exp"` and `[class.foo]="fooExp"` on one element. In the current view engine those things are independent, would like to see explicit test for this situation in ivy.

File: packages/core/test/render3/integration_spec.ts [hunk 5]
```diff
@@ -1517,6 +1518,63 @@ describe('render3 integration test', () => {
...
+           }, 1, 0, [DirWithClassDirective]);
+
+           const fixture = new ComponentFixture(App);
+           expect(mockClassDirective !.classesVal).toEqual('apple orange banana');
+         });
+
+      it('should update `[class]` and bindings in the provided directive if the input is matched',
+         () => {
+           /**
+            * <my-comp DirWithClass></my-comp>
```
Comment (pkozlowski-opensource) [inline packages/core/test/render3/integration_spec.ts]: There should be `[class]` binding somewhere here, right?

File: packages/core/test/render3/integration_spec.ts [hunk 6]
```diff
@@ -8,10 +8,11 @@
 
 import {ElementRef, TemplateRef, ViewContainerRef} from '@angular/core';
 
+import {Input} from '../../src/metadata/directives';
```
Comment (pkozlowski-opensource) [inline packages/core/test/render3/integration_spec.ts]: Seems like we don't need this import. And if we do than we should import it from `@angular/core`

**Comment by pkozlowski-opensource** [comment]
@matsko I've run the existing `NgClass` tests on top of this fix and while it does improve the situation a bit (we've got still 5 tests failing as compared to 7 tests failing without your PR) still the `NgClass` impl is broken with ivy. Broken tests are:
* [should co-operate with the class attribute and binding to it](https://github.com/angular/angular/blob/100c7eff255b43b96753c51875681d1883d8ca61/packages/common/test/directives/ng_class_spec.ts#L13)
* [should co-operate with the interpolated class attribute when interpolation changes](https://github.com/angular/angular/blob/100c7eff255b43b96753c51875681d1883d8ca61/packages/common/test/directives/ng_class_spec.ts#L293)
* [should co-operate with initial class and class attribute binding when binding changes](https://github.com/angular/angular/blob/100c7eff255b43b96753c51875681d1883d8ca61/packages/common/test/directives/ng_class_spec.ts#L337)
* [should co-operate with the class attribute and class.name binding](https://github.com/angular/angular/blob/100c7eff255b43b96753c51875681d1883d8ca61/packages/common/test/directives/ng_class_spec.ts#L319)
* [should co-operate with the interpolated class attribute](https://github.com/angular/angular/blob/100c7eff255b43b96753c51875681d1883d8ca61/packages/common/test/directives/ng_class_spec.ts#L279)

You can see failing tests by adding `fdescribe` to [NgClass tests](https://github.com/angular/angular/blob/100c7eff255b43b96753c51875681d1883d8ca61/packages/common/test/directives/ng_class_spec.ts#L13) and running `bazel test //packages/common/test/... --define=compile=jit
`

File: packages/core/test/render3/integration_spec.ts (hunk 4 above)
Comment (pkozlowski-opensource) [reply packages/core/test/render3/integration_spec.ts]: Analysing `NgClass` tests failures (see https://github.com/angular/angular/pull/26559#issuecomment-431832106) made me realise that we also need a test for `class` interpolation (ex.: `class="foo {{exp}} bar"`.

Also looking at the `NgClass` tests we can see that there so many other, different ways of setting the CSS classes, ex.: `[attr.class]="exp"`.

The existing `NgClass` implementation was _somehow_ making all those crazy cases work together so we either need to replicate this behaviour or design a different system...

**Review by mhevery — APPROVED** [review]
Move `NO_CHANGE` to a `tokens.ts` file.

**Pull request closed without merging**